### DISCUSSION TITLE
Bluetooth for Wifi credential

### DISCUSSION THREAD

**Discussion by flashdato** [pschatzmann/arduino-audio-tools#1381] [Q&A] [answered]
I am building wifi speaker and at this point, I need to be able to write wifi credentials on the board.

I have found this code example BUT I can not integrate it into my Wifi Speaker module

`

#include "AudioTools.h"
#include "AudioCodecs/CodecMP3Helix.h"
#include "AudioLibs/AudioBoardStream.h"
#include "BluetoothSerial.h"  //Header File for Serial Bluetooth, will be added by default into Arduino
#include <EEPROM.h>
#include "WiFi.h"

const char *urls[] = {
  "https://tale-server.onrender.com/stream"
};

ICYStream urlStream(ssid, pass);
AudioSourceURL source(urlStream, urls, "audio/mp3");
AudioBoardStream kit(AudioKitEs8388V1);
MP3DecoderHelix decoder;
AudioPlayer player(source, kit, decoder);

float volume = 1;

//bluetooth parameters

BluetoothSerial ESP_BT;  //Object for Bluetooth
String buffer_in;
unsigned long previousMillis = 0;
byte val;
int addr = 0;
byte indS = 0;
byte indP = 0;
String stream;
byte len = 0;
String temp;
String temp2;
unsigned int interval = 20000;

void next(bool, int, void *) {
  player.next();
}

void up(bool, int, void *) {
  volume = volume + 0.1 > 1 ? 1 : volume + 0.1;
  player.setVolume(volume);
}

void down(bool, int, void *) {
  volume = volume - 0.1 < 0 ? 0 : volume - 0.1;
  player.setVolume(volume);
}

void previous(bool, int, void *) {
  player.previous();
}

void stopResume(bool, int, void *) {
  if (player.isActive()) {
    player.stop();
  } else {
    player.play();
  }
}

// Arduino setup
void setup() {
  EEPROM.begin(50);
  Serial.begin(115200);
  Serial.println("Bluetooth Device is Ready to Pair");

  Serial.println("Waiting For Wifi Updates 60 seconds");
  ESP_BT.begin("Zgaprebis Yuti");  //Name of your Bluetooth Signal

  while (!check_wifiUpdate() == true) {
  }
  Serial.println("The Stored Wifi credetial are : ");
  for (int i = 0; i < 50; i++) {
    val = EEPROM.read(i);
    stream += (char)val;
    if ((val == 10) && (indS == 0)) {
      indS = i;
      Serial.println("indS" + (String)i);
    } else if (val == 10 && indP == 0) {
      indP = i;
      break;
      Serial.println("indP" + (String)i);
    }
  }
  // Serial.println(stream);
  // Serial.println("Stream Ended");
  temp = stream.substring(0, indS);
  temp = temp.substring(5, indS);
  //ssid2=ssid;
  temp2 = stream.substring(indS + 1, indP);
  temp2 = temp2.substring(5, indP - indS);

  int i = temp.length();
  int j = temp2.length();
  char ssid[i];
  char pass[j];
  temp.toCharArray(ssid, i);
  temp2.toCharArray(pass, j);

  Serial.println("Stored SSID");
  Serial.println(ssid);
  Serial.println("Stored PASS");
  Serial.println(pass);
  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, pass);

  if (WiFi.waitForConnectResult() != WL_CONNECTED) {
    Serial.println("WiFi Failed");
    while (1) {
      delay(1000);
    }
  } else {
    Serial.print("Wifi Connected to ");
    Serial.println(ssid);
  }

  //audio

  // setup output
  auto cfg = kit.defaultConfig(TX_MODE);
  cfg.sd_active = false;
  kit.begin(cfg);

  // setup navigation
  kit.addAction(kit.getKey(1), stopResume);
  kit.addAction(kit.getKey(4), next);
  kit.addAction(kit.getKey(3), previous);
  kit.addAction(kit.getKey(5), up);
  kit.addAction(kit.getKey(6), down);

  // setup player
  player.setVolume(1);
  player.begin();
}

boolean check_wifiUpdate() {
  unsigned long currentMillis = millis();

  if (currentMillis - previousMillis >= interval) {
    previousMillis = currentMillis;

    Serial.println("60 Seconds Over");

    return true;
  }

  else if (ESP_BT.available())  //Check if we receive anything from Bluetooth
  {
    interval = 50000;
    buffer_in = ESP_BT.readStringUntil('\n');  //Read what we recevive
                                               //Serial.println("Received:"); Serial.println(buffer_in);
    delay(20);

    if (buffer_in.charAt(0) == 'S') {
      for (int i = 0; i < buffer_in.length(); i++) {
        val = (byte)(buffer_in.charAt(i));
        //Serial.println("val "+val);

        EEPROM.write(addr, val);
        //Serial.println(val);
        addr++;
      }
      //Serial.print("New ");
      //Serial.print(buffer_in);
      EEPROM.write(addr, 10);
      addr++;
      EEPROM.commit();
      ESP_BT.println("SSID Stored");
    }

    else if (buffer_in.charAt(0) == 'P') {
      for (int i = 0; i < buffer_in.length(); i++) {
        val = (byte)(buffer_in.charAt(i));
        //Serial.println("val "+val);
        EEPROM.write(addr, val);
        //Serial.println(val);
        addr++;
      }
      //Serial.print("New ");
      //Serial.print(buffer_in);
      EEPROM.write(addr, 10);
      EEPROM.commit();
      ESP_BT.println("Password Stored");
      return true;
    }

    return false;
  } else {
    return false;
  }
}

// Arduino loop
void loop() {
  player.copy();
  kit.processActions();
}`

this code could be awful so tell me if there is example or some other project on ESP32

**Comment by pschatzmann** [comment, marked as answer]
I don't understand you! 
If you do the authentication outside of the ICYStream or URLStream you just don't provide any login parameters!
I just added this functionality so that I don't have to code the wifi login functionality in each sketch.

You can save some lines of code if you provide the password and ssid via the corresponding methods...

**Reply by flashdato** [reply]
Firstly I want to connect with bluetooth to give board credentials of wifi and save it for future uses 

and after for example 60 seconds I want module to go on with streaming mode on URL provided 

here is only Bluetooth part of the code 

`#include "BluetoothSerial.h"  //Header File for Serial Bluetooth, will be added by default into Arduino
#include <EEPROM.h>
#include "WiFi.h"
BluetoothSerial ESP_BT;  //Object for Bluetooth
String buffer_in;
unsigned long previousMillis = 0;
byte val;
int addr = 0;
byte indS = 0;
byte indP = 0;
String stream;
byte len = 0;
String temp;
String temp2;
unsigned int interval = 60000;
void setup() {
  EEPROM.begin(50);
  Serial.begin(115200);  //Start Serial monitor in 9600
  Serial.println("Bluetooth Device is Ready to Pair");

  Serial.println("Waiting For Wifi Updates 60 seconds");
  ESP_BT.begin("Zgaprebis Yuti");  //Name of your Bluetooth Signal

  while (!check_wifiUpdate() == true) {}
  Serial.println("The Stored Wifi credetial are : ");
  for (int i = 0; i < 50; i++) {
    val = EEPROM.read(i);
    stream += (char)val;
    if ((val == 10) && (indS == 0)) {
      indS = i;
      //Serial.println("indS"+(String)i);
    } else if (val == 10 && indP == 0) {
      indP = i;
      break;
      //Serial.println("indP"+(String)i);
    }
  }
  // Serial.println(stream);
  // Serial.println("Stream Ended");
  temp = stream.substring(0, indS);
  temp = temp.substring(5, indS);
  //ssid2=ssid;
  temp2 = stream.substring(indS + 1, indP);
  temp2 = temp2.substring(5, indP - indS);

  int i = temp.length();
  int j = temp2.length();
  char ssid[i];
  char pass[j];
  temp.toCharArray(ssid, i);
  temp2.toCharArray(pass, j);

  Serial.println("Stored SSID");
  Serial.println(ssid);
  Serial.println("Stored PASS");
  Serial.println(pass);
  WiFi.mode(WIFI_STA);
  WiFi.begin(ssid, pass);

  if (WiFi.waitForConnectResult() != WL_CONNECTED) {
    Serial.println("WiFi Failed");
    delay(1000);
  } else {
    Serial.print("Wifi Connected to ");
    Serial.println(ssid);
  }
}

boolean check_wifiUpdate() {
  unsigned long currentMillis = millis();

  if (currentMillis - previousMillis >= interval) {
    previousMillis = currentMillis;


    Serial.println("60 Seconds Over");

    return true;
  }

  else if (ESP_BT.available())  //Check if we receive anything from Bluetooth
  {
    interval = 50000;
    buffer_in = ESP_BT.readStringUntil('\n');  //Read what we recevive
                                               //Serial.println("Received:"); Serial.println(buffer_in);
    delay(20);


    if (buffer_in.charAt(0) == 'S') {
      for (int i = 0; i < buffer_in.length(); i++) {
        val = (byte)(buffer_in.charAt(i));
        //Serial.println("val "+val);

        EEPROM.write(addr, val);
        //Serial.println(val);
        addr++;
      }
      //Serial.print("New ");
      //Serial.print(buffer_in);
      EEPROM.write(addr, 10);
      addr++;
      EEPROM.commit();
      ESP_BT.println("SSID Stored");
    }

    else if (buffer_in.charAt(0) == 'P') {
      for (int i = 0; i < buffer_in.length(); i++) {
        val = (byte)(buffer_in.charAt(i));
        //Serial.println("val "+val);
        EEPROM.write(addr, val);
        //Serial.println(val);
        addr++;
      }
      //Serial.print("New ");
      //Serial.print(buffer_in);
      EEPROM.write(addr, 10);
      EEPROM.commit();
      ESP_BT.println("Password Stored");
      return true;
    }

    return false;
  } else {
    return false;
  }
}

void loop() {
}
`

**Reply by pschatzmann** [reply]
I am afraid I still don't understand your problem that you have with my library!

**Reply by flashdato** [reply]
Basically I want it to wait for me it is starting before setup and code I am adding needs setup function to get wifi credentials

How can I jump over that basically delay initializing ur library and wifi

**Reply by pschatzmann** [reply]
You can move the logic into the loop

**Reply by flashdato** [reply]
No because than other functions for example volume up wont work because it needs player defined

**Reply by pschatzmann** [reply]
You will need to manage the state and not call any functionality which is not available yet!
Even the login should be in the loop and be executed whenver we are not logged in yet or lost the connection
